File: elixir/tournament/tournament.exs
defmodule Tournament do
  defmodule Stat do
    defstruct mp: 0, w: 0, d: 0, l: 0, p: 0
  end

  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """
  @spec tally(input :: list(String.t())) :: String.t()
  def tally(input) do
    import String, only: [pad_trailing: 2, pad_leading: 2]

    input
    |> Enum.map(&String.split(&1, ";"))
    |> Enum.filter(&(length(&1) == 3 && Enum.at(&1, 2) in ["win", "loss", "draw"]))
    |> Enum.reduce(%{}, fn [team1, team2, outcome], acc ->
      acc
      |> Map.put(team1, stats(Map.get(acc, team1, %Stat{}), outcome))
      |> Map.put(team2, stats(Map.get(acc, team2, %Stat{}), opposite_outcome(outcome)))
    end)
    |> Enum.sort_by(fn {name, hist} -> {-hist.p, name} end)
    |> Enum.reduce("Team                           | MP |  W |  D |  L |  P", fn {name, hist}, acc ->
      acc <>
        "\n" <>
        Enum.join(
          [
            pad_trailing(name, 30),
            pad_leading(to_string(hist.mp), 2),
            pad_leading(to_string(hist.w), 2),
            pad_leading(to_string(hist.d), 2),
            pad_leading(to_string(hist.l), 2),
            pad_leading(to_string(hist.p), 2)
          ],
          " | "
        )
    end)
  end

  defp stats(team, "win"), do: %{team | mp: team.mp + 1, w: team.w + 1, p: team.p + 3}
  defp stats(team, "loss"), do: %{team | mp: team.mp + 1, l: team.l + 1}
  defp stats(team, "draw"), do: %{team | mp: team.mp + 1, d: team.d + 1, p: team.p + 1}

  defp opposite_outcome(outcome) do
    case outcome do
      "win" -> "loss"
      "loss" -> "win"
      "draw" -> "draw"
    end
  end
end

File: csharp/tournament/Tournament.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;

public static class Tournament
{   
    public static void Tally(Stream inStream, Stream outStream)
    {
        var teams = new Dictionary<string, Team>();
        using var reader = new StreamReader(inStream);
        using var writer = new StreamWriter(outStream);

        while (reader.Peek() != -1)
        {
            var split = reader.ReadLine().Split(';');
            var team1 = teams.ContainsKey(split[0]) ? teams[split[0]] : new Team { Name = split[0] };
            var team2 = teams.ContainsKey(split[1]) ? teams[split[1]] : new Team { Name = split[1] };

            switch (split[2])
            {
                case "win":
                    team1.Points += 3;
                    team1.Won++;
                    team2.Lost++;
                    break;
                case "loss":
                    team2.Points += 3;
                    team1.Lost++;
                    team2.Won++;
                    break;
                case "draw":
                    team1.Points++;
                    team2.Points++;
                    team1.Drawn++;
                    team2.Drawn++;
                    break;
            }

            teams[split[0]] = team1;
            teams[split[1]] = team2;
        }

        var output = new List<string>(teams.Count + 1) { "Team                           | MP |  W |  D |  L |  P" };
        output.AddRange(teams.Values.OrderByDescending(t => t.Points).ThenBy(t => t.Name).Select(t => t.PrintRow()));
        writer.Write(string.Join('\n', output));
    }
}

public class Team
{
    public string Name { get; set; }
    public int Played => Won + Drawn + Lost;
    public int Won { get; set; }
    public int Drawn { get; set; }
    public int Lost { get; set; }
    public int Points { get; set; }
    public string PrintRow() =>
        $"{Name,-30} | {Played,2} | {Won,2} | {Drawn,2} | {Lost,2} | {Points,2}";
}